frontend/src/widgets/genericbar.js

```diff
@@ -105,172 +98,181 @@ export class GenericBarWidget extends React.Component {
+              sortOrder={sortOrder}
+              horizontal={horizontal}
+              labels={getLabels()}
               labelComponent={
                 <ChartTooltip
-                  dx={this.props.horizontal ? -10 : 0}
-                  dy={this.props.horizontal ? 0 : -10}
-                  style={{ fill: 'white', fontSize: this.props.fontSize-2 || 14}}
+                  dx={horizontal ? -10 : 0}
+                  dy={horizontal ? 0 : -10}
+                  style={{ fill: 'white', fontSize: fontSize-2 || 14}}
                 />
               }
             />
           );
         }
       }
     }
-    this.setState({barCharts});
-  }
+    setBarCharts(barCharts);
+  }, [data]);
 
-  getChartHeight = (numBars) => {
+  const getChartHeight = (numBars) => {
     if (numBars > 10) {
       return numBars*30;
     }
     else {
       return 300;
     }
-  }
+  };
 
-  componentDidMount() {
-    this.getData();
-  }
-
-  componentDidUpdate(prevProps) {
-    if (prevProps.params !== this.props.params) {
-      this.params = this.props.params;
-      this.getData();
-    }
-  }
-
-  getDropdowns() {
-    if (this.props.hideDropdown) {
+  const getDropdowns = () => {
+    if (hideDropdown) {
       return null;
     }
     else {
       return (
         <div>
           <ParamDropdown
-            dropdownItems={this.props.dropdownItems || ['component', 'env', 'metadata.jenkins.job_name']}
-            defaultValue={this.params.group_field}
-            handleSelect={this.onGroupFieldSelect}
+            dropdownItems={dropdownItems}
+            defaultValue={group_field}
+            handleSelect={onGroupFieldSelect}
             tooltip="Group data by:"
           />
           <ParamDropdown
             dropdownItems={[1, 2, 3, 4, 5, 6]}
-            handleSelect={this.onWeekSelect}
-            defaultValue={this.params.weeks}
+            handleSelect={onWeekSelect}
+            defaultValue={params.weeks}
             tooltip="Set weeks to:"
           />
         </div>
       )
     }
-  }
+  };
 
-  render() {
-    return (
-      <Card>
-        <WidgetHeader title={this.title} getDataFunc={this.getData} onEditClick={this.props.onEditClick} onDeleteClick={this.props.onDeleteClick}/>
-        <CardBody data-id="recent-runs">
-          {this.state.genericBarError &&
-            <p>Error fetching data</p>
-          }
-          {(!this.state.genericBarError && this.state.isLoading) &&
-          <Text component="h2">Loading ...</Text>
-          }
-          {(!this.state.genericBarError && !this.state.isLoading) &&
-          <Chart
-            domainPadding={ this.props.horizontal ? { x: 20 } : { y: 20} }
-            padding={ this.props.padding || {
-              bottom: 30,
-              left: 150,
-              right: 15,
-              top: 20
+  return (
+    <Card>
+      <WidgetHeader title={title} getDataFunc={getData} onEditClick={onEditClick} onDeleteClick={onDeleteClick}/>
```

Comment:
You can drop passing `getDataFunc` for this, I plan to drop it from WidgetHeader.  It was necessary when you couldn't reload the page because there was not static pathing.  Now you can reload the page, instead of trying to individually reload some widget's content.

This allows for making `getData` a `useEffect` function itself.